### ISSUE TITLE
Add samples that use POINTS and LINES and other mesh modes

### ISSUE THREAD

**Issue by emackey** [KhronosGroup/glTF-Sample-Models#124]
A glTF mesh can have a number of different modes:

https://github.com/KhronosGroup/glTF/blob/c02fd64e1fc15985d84cbbfacb536709f7c22e28/specification/2.0/schema/mesh.primitive.schema.json#L32-L56

We should have samples that exercise all of these modes.  In particular, `POINTS` and `LINES` are very different from the other modes, and need their own samples.  ~~But all modes should be tested.~~ (Edit: Perhaps not, see below.)

See KhronosGroup/glTF#917 for upcoming spec clarification on lighting of points and lines.

**Comment by bghgary** [comment]
The asset generator project will generate unit models with an asset for each primitive modes. See our roadmap here: https://github.com/bghgary/glTF-Asset-Generator/issues/63.

**Comment by donmccurdy** [comment]
There is also an "Outlined Box" example donated to three.js: https://threejs.org/examples/?q=gltf#webgl_loader_gltf

**Comment by andreasplesch** [comment]
see also #106

**Comment by bghgary** [comment]
These are basically dupes. Should we merge the threads? We are very close to releasing a version of generated assets that has all of these modes.

**Comment by andreasplesch** [comment]
I closed #106 now that the issues are cross-referenced.

**Comment by bghgary** [comment]
v0.4 of the generated assets now has all of these modes: https://github.com/bghgary/glTF-Asset-Generator/blob/v0.4.0/Output/Mesh_Indices/README.md

**Comment by javagl** [comment]
We should probably think about how to align the models from this repo with the models from the asset generator. While the requirement for having models with different modes is covered with the asset generator output, I wonder whether this issue could be closed, or whether we should explicitly add _one_ model _here_ that shows the different rendering modes at once.  (I think there should be such a model, as one of the `Simple....` test models, and maybe I'll try to allocate some time to create one...)

**Comment by donmccurdy** [comment]
Note that Blender can export these too, if you dissolve faces and export with the "Loose Edges" and "Loose Points" options. Could be helpful if creating more than primitive shapes.

**Comment by javagl** [comment]
I considered https://github.com/KhronosGroup/glTF-Sample-Models/pull/329 to eventually cover this issue to _some_ extent (and that sample might have been created with blender...?). But the focus there is on the rendered appearance with PBR, which is different from what would be the goal of a `Simple...` test model. The latter would really just be a bunch of hexagons, with `GL_POINTS`, `GL_LINES`, `GL_LINE_LOOP`, `GL_LINE_STRIP`, `GL_TRIANGLES`, `GL_TRIANGLE_STRIP` and `GL_TRIANGLE_FAN`, and some info about the differences in the indexing in the README.md.

**Comment by donmccurdy** [comment]
Beyond the scope of a sample model here, but just for fun:

- https://twitter.com/ra/status/1141885099044233216
- https://fern-wildebeest.glitch.me/

The idea is to start with GL_LINES, and do post-processing for the intentionally pixelated effect.

**Comment by emackey** [comment]
I've recently wondered if `GL_TRIANGLE_STRIP` and `GL_TRIANGLE_FAN` are ever used in the wild.  Back in the late 90's ~ early 00's, one of my then-coworkers became enamored with triangle strips and I've recently been taking a close look at a large number of old models that came through his converter.  Even a relatively simple mesh of a few thousand triangles typically requires dozens or hundreds of separate triangle strips.  A "cylinder" primitive in this old format used one triangle strip for the walls and two separate triangle fans as the end caps.  These would be separate draw calls, or separate glTF primitives, right?  Can a glTF `GL_TRIANGLE_STRIP` primitive contain more than one strip?  That would require Primitive Restart which glTF doesn't offer, right?

In my efforts to upgrade these old models to glTF, I've arrived at the conclusion that undoing all of the triangle strip/fan work and throwing all the triangles into a single primitive seems better.  Is there any reason to favor strips/fans?

Certainly this sample repo should have `GL_LINES` and `GL_POINTS` (separately I would think, for debugging and conformance testing).  I'm not so sure we should be promoting strips and fans.

**Comment by donmccurdy** [comment]
We (three.js) have already removed support for GL_TRIANGLE_STRIP and GL_TRIANGLE_FAN. glTF files containing those primitives are converted at load time, but in practice I've never seen this feature used.

Related discussion:

- https://github.com/mrdoob/three.js/pull/18041

***

Totally defer to maintainers on what an appropriate GL_LINES sample would be, but here's a model I built in Blender. Loosely inspired by ruined temples from the game, [Gris](https://www.nintendo.com/store/products/gris-switch/). Tested in three.js and Babylon.js.

[TempleLines.glb.zip](https://github.com/KhronosGroup/glTF-Sample-Models/files/10441160/TempleLines.glb.zip)

![Screenshot 2023-01-17 at 8 51 56 PM](https://user-images.githubusercontent.com/1848368/213062610-0b03575f-580b-48fd-ad60-5627320c7cca.png)

**Comment by michaliskambi** [comment]
I found this thread when looking for glTF test models using lines and points.

- For testing Castle Game Engine, we have 2 testcases: https://github.com/castle-engine/demo-models/tree/master/gltf/lines , https://github.com/castle-engine/demo-models/tree/master/gltf/points . Exported from Blender. They are extremely ugly and simple, but they exist :), with screenshots, so if anyone else finds this thread looking for test glTF models -- you're welcome to use them :)

- I also tested Castle Game Engine with temple model above from @donmccurdy , thanks! Used a screenshot of it in https://castle-engine.io/wp/2023/02/24/rendering-of-lines-and-points-from-gltf-more-options-for-line-nodes-in-x3d/ .
